src/mame/drivers/seta.cpp

```diff
@@ -3026,16 +3026,90 @@ void kiwame_state::kiwame_map(address_map &map)
+
+   A2  -> I2
+   A3  -> I3
+   A6  -> I4
+   A8  -> I5
+   A11 -> I6
+   A13 -> I7
+   A15 -> I8
+   A16 -> I9
+
+   Data sent by the CPU during these writting operations is not used to calculate the protection register value, only some
+   address lines are used to compute the value, as shown above.
+   Every write operation done to the PAL discards the previous stored value in the registers and stores a new computed value,
+   following the logic equations programmed in the PAL.
+
+   (I1 = CLK : pulses when accessing to writting handler and acts as clock for internal latches in PAL)
+   (I11 = /OE : asserted when accessing to reading handler. Sub-address used for reading here is no used to compute the result value
+
+   I19  -> D0
+   I18  -> D1
+   I17  -> D2
+   I16  -> D3
+   I15  -> D4
+   I14  -> D5
+   I13  -> D6
+   I12  -> D7
+*/
 u16 seta_state::thunderl_protection_r()
 {
 //  logerror("PC %06X - Protection Read\n", m_maincpu->pc());
-	return 0x00dd;
-}
-void seta_state::thunderl_protection_w(u16 data)
-{
-//  logerror("PC %06X - Protection Written: %04X <- %04X\n", m_maincpu->pc(), offset * 2, data);
+
+	return m_thunderl_protection_reg;
+}
+void seta_state::thunderl_protection_w(offs_t offset, u16 data)
+{
+	const u32 addr = offset * 2;
+	// data byte written here is not used to save the value into the protection register
+	/*
+	const u8 init_prot_data =
+		((addr & 0x04) >> 2)                       // A2 (addr & 2^address_bit) >> (address_bit - current_data_bit)
+		| ((addr & 0x8) >> 2)                      // A3
+		| ((addr & 0x40) >> 4)                     // A6 
+		| ((addr & 0x100) >> 5)                    // A8
+		| ((addr & 0x800) >> 7)                    // A11
+		| ((addr & 0x2000) >> 8)                   // A13
+		| ((addr & 0x8000) >> 9)                   // A15
+		| ((addr & 0x10000) >> 9);                 // A16
+	*/
+
+	m_thunderl_protection_reg =
+		((addr & 0x04) >> 2)                       // D0 = A2
+		| (((addr & 0x04) >> 1)                    // D1 = A2 & /A3
+			& ((~addr & 0x8) >> 2))
+		| ((addr & 0x04)                           // D2 = A2 | /A6
+			| ((~addr & 0x40) >> 4))
+		| (((addr & 0x04) << 1)                    // D3 = A2 | /A6 | /A8
+			| ((~addr & 0x40) >> 3)
+			| ((~addr & 0x100) >> 5))
+		| (((addr & 0x8) << 1)                     // D4 = A3 & /A11 & A15
+			& ((~addr & 0x800) >> 7)
+			& ((addr & 0x8000) >> 11))
+		| (((addr & 0x40) >> 1)                    // D5 = A6 & A13
+			& ((addr & 0x2000) >> 8))
+		| (((addr & 0x40)                          // D6 = (A6 | /A16) & (A13 | /A16)
+				| ((~addr & 0x10000) >> 10))
+			& (((addr & 0x2000) >> 7)
+				| ((~addr & 0x10000) >> 10)))
+		| ((((addr & 0x40) << 1)                   // D7 = (A6 | /A16) & (A13 | /A16) & (A2 | /A6 | /A8)
+				| ((~addr & 0x10000) >> 9))
+			& (((addr & 0x2000) >> 6)
+				| ((~addr & 0x10000) >> 9))
+			& (((addr & 0x04) << 6)
+				| ((~addr & 0x40) << 1)
+				| ((~addr & 0x100) >> 1)));
```

Comment:
It’s more difficult to work check this for correctness in this form.  It would be a lot more readable if you used the `BIT` helper and just shifted the results into position, something like:

```
	m_thunderl_protection_reg =
			(BIT(addr, 2) << 0)
			| ((BIT(addr, 2) & BIT(~addr, 3)) << 1)
			| ((BIT(addr, 2) | BIT(~addr, 6)) << 2)
			| ((BIT(addr, 2) & BIT(~addr, 6) & BIT(~addr, 8)) << 3)
			| ((BIT(addr, 3) & BIT(~addr, 11) & BIT(addr, 15)) << 4)
			| ((BIT(addr, 6) & BIT(addr, 13)) << 5)
			| (((BIT(addr, 6) & BIT(addr, 13)) | BIT(~addr, 16)) << 6)
			| ((((BIT(addr, 6) & BIT(addr, 13)) | BIT(~addr, 16)) & (BIT(addr, 2) | BIT(~addr, 6) | BIT(~addr, 8))) << 7);
```
If you do it like that, the address bits being used are obvious, and the code matches the comments.  (I haven’t actually tested that code, I just wrote it based on the comments.)